text/0000-lwc-signals.md

```diff
@@ -0,0 +1,373 @@
+1. __Simplified State Management__: Signals provide a straightforward way to manage reactive data. By deprecating the @wire and @track decorators, we can reduce complexity, making it easier for developers to write and maintain component logic.
+1. __Performance Optimizations__: The proposed model opens up possibilities for performance optimizations by allowing selective rehydration of the UI. Components could update the DOM directly in response to state changes without a full re-render, leading to faster updates and reduced resource consumption.
+
+The motivation for this RFC is to align LWC with modern JavaScript practices, simplify component development, and improve performance, making LWC a more attractive option for developers and increasing its competitiveness in the frontend framework market.
+
+## Detailed Design
+
+The detailed design of the new reactivity model revolves around the concept of Signals, which are reactive primitives that allow for fine-grained control over state management and UI updates. Here's how it will look in practice:
+
+### Reactivity with Signals
+
+A Signal is an object that holds a value and allows components to react to changes to that value. It exposes a `.value` property for accessing the current value, and `.subscribe` and `.unsubscribe` methods for responding to changes.
+
+```javascript
+import { signal } from 'signals';
+
+export default class ExampleComponent extends LightningElement {
+    count = signal(0);
+
+    increment() {
+        this.count.value++;
+    }
+}
+```
+
+In the template, we can bind directly to the `.value` property:
+
+```html
+<template>
+    <button onclick={increment}>Increment</button>
+    <p>{count.value}</p>
+</template>
+```
+
+### Compiler Enhancements
+
+The LWC compiler will be updated to recognize when a Signal's `.value` is used within a template. Instead of accessing the property directly at runtime, it will generate code to subscribe to the Signal's changes during component's template initialization and unsubscribe upon component's template disposal.
+
+### Store Protocol
+
+A store must implement the following protocol to be compatible with the new reactivity model:
+
+```javascript
+class Store {
+    get value() { /* return current value */ }
+    subscribe(callback) { /* subscribe to changes */ }
+    unsubscribe(callback) { /* unsubscribe from changes */ }
+}
+```
+
+__NOTE:__ _We can work the details of this protocol. It could be more simple, or hopefully similar to what is out there._
+
+### Integration with Components
+
+Components should treat signals as internal reactive state mechanisms. While a signal can technically be passed down to child components, it's recommended to pass only the necessary data, typically the `.value` of the signal, to keep the child components decoupled from the parent's state management strategy.
```

Comment:
I understand & agree with the sentiment here, but I think injecting a signal into a component is going to be the key to making low-code scenarios work?

Reply:
@caridy I think I remember a demo when signals were introduced for either Solid or React, where signals were explicitly used to skip levels inside component hierarchies (kind of like a poor-man's state management), but where this was key to some major performance improvements observed. Using the following technique there lead to signals "magically" updating their values inside the HTML elements directly, without relying/executing a full re-render of the involved components. I know that this would scare the hell out of you, but that approach was like what brought "lightning" to the demo back then.

```js
// bundle `acme/signals`
import { signal } from 'signals';

export const count = signal(0);
```

```js
// bundle `acme/parent`
import { count } from 'acme/signals';

class Parent extends LightningElement {
    handleClick() {
         count.value++;
    }
}
```

```js
// bundle `acme/grandGrandChild`
import { count } from 'acme/signals';

class GrandGrandChild extends LightningElement {
    get count() {
        return count;
    }
}
```

```html
<!-- bundle `acme/grandGrandChild` -->
<div>{count.value}</div>
```

Reply:
The only thing to really try to figure out is whether or not we should support `{count}` vs `{count.value}` in the templates, a la-vue. The hesitation here is the signal down to other components, the ambiguity he being:

```
<template>
   <input value={count} />
   <x-foo count={count}></x-foo>
</template>
```

Should `count` property on foo receive a signal or its value? That's the problem.

Reply:
> it's recommended to pass only the necessary data

Let's assume for a second that our users won't always do the "recommended" thing. (A very good assumption!)

- What kinds of guardrails can we put in place? Warnings/errors/etc.?
- What happens if they mess it up? What are the consequences?
- Knowing all this, _should_ we make it possible to do the wrong thing?

> Using the following technique there lead to signals "magically" updating their values inside the HTML elements directly, without relying/executing a full re-render of the involved components.

This is called fine-grained reactivity. It is definitely a perf boost, but I'm not convinced that signals are required to achieve it: https://github.com/salesforce/lwc/issues/3624